optimade/filtertransformers/mongo.py

```diff
@@ -81,20 +77,58 @@ def expression_phrase(self, arg):
         # expression_phrase: [ NOT ] ( comparison | "(" expression ")" )
         return self._recursive_expression_phrase(arg)
 
-    def property_first_comparison(self, arg):
+    @v_args(inline=True)
+    def property_first_comparison(self, quantity, query):
         # property_first_comparison: property ( value_op_rhs | known_op_rhs | fuzzy_string_op_rhs | set_op_rhs |
         # set_zip_op_rhs | length_op_rhs )
 
         # Awkwardly, MongoDB will match null fields in $ne filters,
         # so we need to add a check for null equality in evey $ne query.
-        if "$ne" in arg[1]:
-            return {"$and": [{arg[0]: arg[1]}, {arg[0]: {"$ne": None}}]}
-
-        return {arg[0]: arg[1]}
+        if "$ne" in query:
+            return {"$and": [{quantity: query}, {quantity: {"$ne": None}}]}
+
+        # Check if a $size query is being made (indicating a length_op_rhs filter); if so, check for
+        # a defined length alias to replace the $size call with the corresponding filter on the
+        # length quantity then carefully merge the two queries.
+        #
+        # e.g. `("elements", {"$size": 2, "$all": ["Ag", "Au"]})` should become
+        # `{"elements": {"$all": ["Ag", "Au"]}, "nelements": 2}` if the `elements` -> `nelements`
+        # length alias is defined.
+        if "$size" in query:
+            if (
+                getattr(self.backend_mapping.get(quantity), "length_quantity", None)
```

Comment:
This is a bit fiddly so I have double-checked that this and L103 are correct (and that this is properly tested). At this point in the transformation (for mongo at least), `quantity` has already been unwrapped to the backend-specific field name. This is because the aliasing is handled at the innermost level within the `property` rule.

Reply:
Added a test for this: length query on an aliased field that has no defined length alias:

https://github.com/Materials-Consortia/optimade-python-tools/pull/797/commits/94bf36866df6aa2f8ee15f62880269c1801e797f#diff-9c452e6fa97bbc68a7d59adf07c58fbdde726a86c397077f723af22e918ea63fR508-R540